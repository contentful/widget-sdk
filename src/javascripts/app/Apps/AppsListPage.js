import React from 'react';
import PropTypes from 'prop-types';
import AppsFrameworkIntroBanner from './AppsFrameworkIntroBanner';
import { css } from 'emotion';
import { get, partition } from 'lodash';

import tokens from '@contentful/forma-36-tokens';
import {
  SkeletonContainer,
  SkeletonDisplayText,
  SkeletonText,
  SkeletonImage,
  Notification,
  Heading,
  Note,
  TextLink,
  Card,
  Paragraph,
  Workbench,
} from '@contentful/forma-36-react-components';

import NavigationIcon from 'ui/Components/NavigationIcon';

import DocumentTitle from 'components/shared/DocumentTitle';
import ModalLauncher from 'app/common/ModalLauncher';
import FeedbackButton from 'app/common/FeedbackButton';

import { websiteUrl } from 'Config';

import AppListItem from './AppListItem';
import AppDetailsModal from './AppDetailsModal';
import * as AppLifecycleTracking from './AppLifecycleTracking';
import { isUsageExceeded } from './isUsageExceeded';
import { buildUrlWithUtmParams } from 'utils/utmBuilder';

const styles = {
  pricingInfo: css({
    marginBottom: tokens.spacingL,
    zIndex: 3,
  }),
  workbench: css({
    backgroundColor: tokens.colorElementLightest,
  }),
  appListCard: css({
    position: 'relative',
    marginBottom: tokens.spacingL,
  }),
  overlay: css({
    position: 'absolute',
    zIndex: 2,
    top: 0,
    left: 0,
    width: '100%',
    height: '100%',
    backgroundColor: tokens.colorWhite,
    opacity: 0.8,
  }),
  installedList: css({
    marginBottom: tokens.spacing2Xl,
  }),
};

const externalLinkProps = {
  target: '_blank',
  rel: 'noopener noreferrer',
};

const withInAppHelpUtmParams = buildUrlWithUtmParams({
  source: 'webapp',
  medium: 'pricing-info',
  campaign: 'in-app-help',
});

const openDetailModal = ({ spaceInformation, usageExceeded, canManageApps }) => (app) => {
  AppLifecycleTracking.detailsOpened(app.id);

  ModalLauncher.open(({ isShown, onClose }) => (
    <AppDetailsModal
      isShown={isShown}
      onClose={onClose}
      app={app}
      spaceInformation={spaceInformation}
      usageExceeded={usageExceeded}
      canManageApps={canManageApps}
    />
  ));
};

const Header = () => <Heading>Apps</Heading>;

const PricingInfo = () => (
  <Note
    className={styles.pricingInfo}
    noteType="warning"
    title="Upgrade to our new pricing model to access this feature"
    testId="apps-pricing-info">
    <Paragraph>
      To access this feature, you need to move to the latest version of Spaces. Submit a{' '}
      <TextLink
        href={withInAppHelpUtmParams(websiteUrl('/support/?upgrade-pricing=true'))}
        {...externalLinkProps}>
        support request
      </TextLink>{' '}
      to get started, or learn more about our{' '}
      <TextLink
        href={withInAppHelpUtmParams(
          websiteUrl(
            '/pricing/?faq_category=payments-subscriptions&faq=what-type-of-spaces-can-i-have'
          )
        )}
        {...externalLinkProps}>
        Space types and pricing
      </TextLink>
      .
    </Paragraph>
  </Note>
);

const AppsListShell = (props) => (
  <Workbench className={styles.workbench}>
    <Workbench.Header
      title={<Header />}
      icon={<NavigationIcon icon="apps" color="green" size="large" />}
      actions={<FeedbackButton target="extensibility" about="Apps" label="Give your feedback" />}
    />
    <Workbench.Content type="text">
      <AppsFrameworkIntroBanner canManageApps={props.canManageApps} />
      {props.appsFeatureDisabled && <PricingInfo />}
      <Card padding="large" className={styles.appListCard}>
        {props.appsFeatureDisabled && (
          <div className={styles.overlay} data-test-id="disabled-beta-apps" />
        )}
        <div>{props.children}</div>
      </Card>
    </Workbench.Content>
  </Workbench>
);

AppsListShell.propTypes = {
  appsFeatureDisabled: PropTypes.bool,
  canManageApps: PropTypes.bool,
};

const ItemSkeleton = (props) => (
  <React.Fragment>
    <SkeletonImage offsetTop={props.baseTop} width={36} height={36} radiusX={36} radiusY={36} />
    <SkeletonText offsetTop={props.baseTop + 15} offsetLeft={50} lineHeight={8} width={240} />
    <SkeletonText offsetTop={props.baseTop + 15} offsetLeft={510} lineHeight={8} width={90} />
  </React.Fragment>
);
ItemSkeleton.propTypes = {
  baseTop: PropTypes.number,
};

const AppsListPageLoading = () => {
  return (
    <AppsListShell>
      <SkeletonContainer svgWidth={600} svgHeight={200} ariaLabel="Loading apps list...">
        <SkeletonDisplayText />
        <ItemSkeleton baseTop={60} />
        <ItemSkeleton baseTop={110} />
        <ItemSkeleton baseTop={160} />
      </SkeletonContainer>
    </AppsListShell>
  );
};

export default class AppsListPage extends React.Component {
  static propTypes = {
    goToContent: PropTypes.func.isRequired,
    repo: PropTypes.shape({
      getApps: PropTypes.func.isRequired,
    }).isRequired,
    organizationId: PropTypes.string.isRequired,
    spaceInformation: PropTypes.shape({
      spaceId: PropTypes.string.isRequired,
      spaceName: PropTypes.string.isRequired,
      envMeta: PropTypes.shape({
        environmentId: PropTypes.string.isRequired,
        isMasterEnvironment: PropTypes.bool.isRequired,
        aliasId: PropTypes.string,
      }),
    }).isRequired,
    userId: PropTypes.string.isRequired,
    hasAppsFeature: PropTypes.bool.isRequired,
    deeplinkAppId: PropTypes.string,
    deeplinkReferrer: PropTypes.string,
    canManageApps: PropTypes.bool,
  };

  state = { ready: false };

  async componentDidMount() {
    try {
      const apps = await this.props.repo.getApps();
      const [installedApps, availableApps] = partition(apps, (app) => !!app.appInstallation);

      this.setState({ ready: true, availableApps, installedApps }, () => {
        this.openDeeplinkedAppDetails();
      });
    } catch (err) {
      Notification.error('Failed to load apps.');
    }
  }

  openDeeplinkedAppDetails() {
    const { deeplinkAppId, hasAppsFeature, spaceInformation, canManageApps } = this.props;

    if (!hasAppsFeature || !deeplinkAppId) {
      return;
    }

    const { installedApps, availableApps } = this.state;

    const deeplinkedApp = installedApps.concat(availableApps).find((app) => {
      // Find either by marketplace ID ("slug", pretty)
      // or definition ID (Contentful UUID, ugly).
      const byMarketplaceId = app.id === deeplinkAppId;
      const definitionId = get(app, ['appDefinition', 'sys', 'id']);
      const byDefinitionId = definitionId === deeplinkAppId;

      return byMarketplaceId || byDefinitionId;
    });

    if (deeplinkedApp && !deeplinkedApp.isPrivateApp) {
      // TODO: we could potentially track the deeplink.
      // Use `this.props.deeplinkReferrer`.
      openDetailModal({
        spaceInformation,
        usageExceeded: isUsageExceeded(installedApps),
        canManageApps,
      })(deeplinkedApp);
    }
  }

  render() {
    return (
      <>
        <DocumentTitle title="Apps" />
        {this.state.ready ? this.renderList() : <AppsListPageLoading />}
      </>
    );
  }

  renderList() {
    const { organizationId, spaceInformation, userId, hasAppsFeature, canManageApps } = this.props;
    const { installedApps, availableApps } = this.state;
    const { spaceId } = spaceInformation;
    const usageExceeded = isUsageExceeded(installedApps);

    return (
      <AppsListShell
        organizationId={organizationId}
        spaceId={spaceId}
        userId={userId}
        canManageApps={canManageApps}
        appsFeatureDisabled={!hasAppsFeature}>
        {installedApps.length > 0 && (
          <>
            <Heading element="h2">Installed</Heading>
            <div data-test-id="installed-list" className={styles.installedList}>
              {installedApps.map((app) => (
                <AppListItem
                  key={app.id}
                  app={app}
                  canManageApps={canManageApps}
                  openDetailModal={openDetailModal({ spaceInformation, canManageApps })}
                />
              ))}
            </div>
          </>
        )}
        <Heading element="h2">Available</Heading>
        {availableApps.length > 0 && (
          <div>
            {availableApps.map((app) => (
              <AppListItem
                key={app.id}
                app={app}
                canManageApps={canManageApps}
                openDetailModal={openDetailModal({
                  spaceInformation,
                  usageExceeded,
                  canManageApps,
                })}
              />
            ))}
          </div>
        )}
      </AppsListShell>
    );
  }
}
