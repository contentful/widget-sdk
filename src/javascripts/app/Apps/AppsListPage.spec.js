import React from 'react';
import { render, wait } from '@testing-library/react';
import AppsListPage from './AppsListPage';
import repoAppsMock from './__mocks__/repoAppsMock.json';

jest.mock('app/common/MarkdownRenderer', () => () => null);

jest.mock('MicroBackendsClient', () => {
  return () => {};
});

jest.mock('access_control/AccessChecker', () => ({
  getSectionVisibility: () => ({ apps: true })
}));

describe('AppsListPage', () => {
  const orgId = '123';
  const spaceId = '456';
  const userId = '000';
  const spaceInformation = {
    spaceId: spaceId,
    spaceName: 'my-test-space',
    envMeta: {
      environmentId: 'master',
      isMasterEnvironment: true,
      aliasId: undefined
    }
  };

  it('should match snapshot for loading state', async () => {
    const { container } = render(
      <AppsListPage
        goToContent={() => {}}
        repo={{ getApps: jest.fn() }}
        organizationId={orgId}
        spaceId={spaceId}
        userId={userId}
        hasAppsFeature={true}
        spaceInformation={spaceInformation}
        canManageApps={true}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot of apps loaded state', async () => {
    const mockRepo = {
      getApps: jest.fn(() => Promise.resolve(repoAppsMock))
    };

    const { container } = render(
      <AppsListPage
        goToContent={() => {}}
        repo={mockRepo}
        organizationId={orgId}
        spaceId={spaceId}
        userId={userId}
        hasAppsFeature={true}
        spaceInformation={spaceInformation}
        canManageApps={true}
      />
    );

    await wait();

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot of apps loaded with limited access', async () => {
    const mockRepo = {
      getApps: jest.fn(() => Promise.resolve(repoAppsMock))
    };

    const { container } = render(
      <AppsListPage
        goToContent={() => {}}
        repo={mockRepo}
        organizationId={orgId}
        spaceId={spaceId}
        userId={userId}
        hasAppsFeature={true}
        spaceInformation={spaceInformation}
        // limit the access
        canManageApps={false}
      />
    );

    await wait();

    expect(container).toMatchSnapshot();
  });
});
