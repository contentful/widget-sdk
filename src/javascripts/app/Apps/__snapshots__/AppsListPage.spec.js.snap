// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppsListPage should match snapshot for loading state 1`] = `
.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 1.5rem;
}

.emotion-2 {
  margin-top: 1rem;
}

.emotion-3 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.emotion-0 {
  font-size: 1rem;
  margin-bottom: 0.75rem;
}

.emotion-1 {
  font-size: 0.875rem;
  margin-bottom: 0.75rem;
}

.emotion-5 {
  margin-bottom: 1.5rem;
}

.emotion-7 {
  background-color: #f7f9fa;
}

.emotion-6 {
  position: relative;
  margin-bottom: 1.5rem;
}

<div>
  <div
    class="Workbench__Workbench___THi9K emotion-7"
    data-test-id="cf-ui-workbench"
  >
    <div
      class="Workbench__Workbench__header___2wo2E"
      data-test-id="cf-ui-workbench-header"
    >
      <div
        class="Workbench__Workbench__header-icon___1kJlL"
      >
        <span
          class="icon-component"
          data-icon-name="page-apps"
        >
          <page-apps.svg />
        </span>
      </div>
      <div
        class="Workbench__Workbench__header-title___3xp7b"
        data-test-id="workbench-title"
      >
        <h1
          class="Heading__Heading___83D3K"
          data-test-id="cf-ui-heading"
        >
          Apps
        </h1>
      </div>
      <div
        class="Workbench__Workbench__header-actions___mJoSR"
      >
        <button
          class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
          data-test-id="cf-ui-text-link"
          type="button"
        >
          <span
            class="TabFocusTrap__TabFocusTrap___39Vty"
            tabindex="-1"
          >
            Give your feedback
          </span>
        </button>
      </div>
    </div>
    <div
      class="Workbench__Workbench__content-wrapper___3yWKi"
    >
      <div
        class="Workbench__Workbench__content___1U8bV"
        data-test-id="cf-ui-workbench-content"
      >
        <div
          class="Workbench__Workbench__content-inner___36_Hx Workbench__Workbench__content-inner--text___7nfTX"
        >
          <div
            class="Card__Card___xXJsT emotion-4 Card__Card--padding-large___3Fgal"
            data-test-id="cf-ui-card"
          >
            <div>
              <h1
                class="Heading__Heading___83D3K emotion-0"
                data-test-id="cf-ui-heading"
              >
                Introducing Contentful Apps
              </h1>
              <p
                class="Paragraph__Paragraph___2aO-9 emotion-1"
                data-test-id="cf-ui-paragraph"
              >
                Contentful apps extend the capabilities of the Contentful web app and the editors who use it. Apps empower you to integrate third-party services, build efficient workflows and customize the functionality of the Contentful web app.
              </p>
              <a
                class="Button__Button___1ZfFj a11y__focus-border--default___60AXp emotion-2 Button__Button--primary___JImeO"
                data-test-id="cf-ui-button"
                href="https://www.contentful.com/developers/docs/extensibility/app-framework/"
                rel="noopener noreferrer"
                target="_blank"
                type="button"
              >
                <span
                  class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
                  tabindex="-1"
                >
                  <span
                    class="Button__Button__label___3tcOj"
                  >
                    Learn how to build your own app
                  </span>
                </span>
              </a>
            </div>
            <div
              class="emotion-3"
            >
              <connected-shapes.svg
                classname="css-1z01nmi-svgContainerExtension"
              />
            </div>
          </div>
          <div
            class="Card__Card___xXJsT emotion-6 Card__Card--padding-large___3Fgal"
            data-test-id="cf-ui-card"
          >
            <p
              class="Paragraph__Paragraph___2aO-9 emotion-5"
              data-test-id="cf-ui-paragraph"
            >
              Contentful apps extend and expand the capabilities of the Contentful web app. You can integrate your favorite third-party services, build better workflows and customize what you can do with Contentful.
            </p>
            <div>
              <svg
                aria-label="Loading apps list..."
                class="SkeletonContainer__SkeletonContainer___23jiu"
                data-test-id="cf-ui-skeleton-form"
                height="200"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                width="600"
              >
                <title>
                  Loading apps list...
                </title>
                <rect
                  clip-path="url(#cf-ui-skeleton-clip-id)"
                  height="100%"
                  style="fill: url(#cf-ui-skeleton-clip-gradient);"
                  width="100%"
                  x="0"
                  y="0"
                />
                <defs>
                  <clippath
                    id="cf-ui-skeleton-clip-id"
                  >
                    <rect
                      height="21"
                      rx="0"
                      ry="0"
                      width="100"
                      x="0"
                      y="0"
                    />
                    <rect
                      height="36"
                      rx="36"
                      ry="36"
                      width="36"
                      y="60"
                    />
                    <rect
                      height="8"
                      rx="0"
                      ry="0"
                      width="240"
                      x="50"
                      y="75"
                    />
                    <rect
                      height="8"
                      rx="0"
                      ry="0"
                      width="90"
                      x="510"
                      y="75"
                    />
                    <rect
                      height="36"
                      rx="36"
                      ry="36"
                      width="36"
                      y="110"
                    />
                    <rect
                      height="8"
                      rx="0"
                      ry="0"
                      width="240"
                      x="50"
                      y="125"
                    />
                    <rect
                      height="8"
                      rx="0"
                      ry="0"
                      width="90"
                      x="510"
                      y="125"
                    />
                    <rect
                      height="36"
                      rx="36"
                      ry="36"
                      width="36"
                      y="160"
                    />
                    <rect
                      height="8"
                      rx="0"
                      ry="0"
                      width="240"
                      x="50"
                      y="175"
                    />
                    <rect
                      height="8"
                      rx="0"
                      ry="0"
                      width="90"
                      x="510"
                      y="175"
                    />
                  </clippath>
                  <lineargradient
                    id="cf-ui-skeleton-clip-gradient"
                  >
                    <stop
                      offset="0%"
                      stop-color="#e5ebed"
                      stop-opacity="1"
                    >
                      <animate
                        attributeName="stop-color"
                        dur="2s"
                        id="animation1"
                        repeatCount="indefinite"
                        values="#e5ebed; #f7f9fa; #e5ebed"
                      />
                    </stop>
                    <stop
                      offset="50%"
                      stop-color="#f7f9fa"
                      stop-opacity="1"
                    >
                      <animate
                        attributeName="stop-color"
                        begin="animation1.begin+0.25s"
                        dur="2s"
                        repeatCount="indefinite"
                        values="#e5ebed; #f7f9fa; #e5ebed"
                      />
                    </stop>
                    <stop
                      offset="100%"
                      stop-color="#e5ebed"
                      stop-opacity="1"
                    >
                      <animate
                        attributeName="stop-color"
                        begin="animation1.begin+0.5s"
                        dur="2s"
                        repeatCount="indefinite"
                        values="#e5ebed; #f7f9fa; #e5ebed"
                      />
                    </stop>
                  </lineargradient>
                </defs>
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AppsListPage should match snapshot of apps loaded state 1`] = `
.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 1.5rem;
}

.emotion-2 {
  margin-top: 1rem;
}

.emotion-3 {
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.emotion-0 {
  font-size: 1rem;
  margin-bottom: 0.75rem;
}

.emotion-1 {
  font-size: 0.875rem;
  margin-bottom: 0.75rem;
}

.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem 0;
  border-bottom: 1px solid #e5ebed;
}

.emotion-12:last-child {
  border-bottom: none;
  padding-bottom: 0;
}

.emotion-10 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  display: block;
}

.emotion-7 {
  font-size: 0.75rem;
  color: #b4c3ca;
}

.emotion-9 {
  font-size: 1rem;
  font-weight: 500;
}

.emotion-11 {
  display: block;
}

.emotion-11 button {
  margin-left: 1rem;
}

.emotion-6 {
  border-radius: 5px;
  background-color: #fff;
  box-shadow: 0 2px 3px 0 rgba(0,0,0,0.08);
  padding: 2px;
  margin-right: 0.75rem;
  width: 35px;
  height: 35px;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.emotion-5 {
  margin-bottom: 1.5rem;
}

.emotion-49 {
  background-color: #f7f9fa;
}

.emotion-48 {
  position: relative;
  margin-bottom: 1.5rem;
}

<div>
  <div
    class="Workbench__Workbench___THi9K emotion-49"
    data-test-id="cf-ui-workbench"
  >
    <div
      class="Workbench__Workbench__header___2wo2E"
      data-test-id="cf-ui-workbench-header"
    >
      <div
        class="Workbench__Workbench__header-icon___1kJlL"
      >
        <span
          class="icon-component"
          data-icon-name="page-apps"
        >
          <page-apps.svg />
        </span>
      </div>
      <div
        class="Workbench__Workbench__header-title___3xp7b"
        data-test-id="workbench-title"
      >
        <h1
          class="Heading__Heading___83D3K"
          data-test-id="cf-ui-heading"
        >
          Apps
        </h1>
      </div>
      <div
        class="Workbench__Workbench__header-actions___mJoSR"
      >
        <button
          class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
          data-test-id="cf-ui-text-link"
          type="button"
        >
          <span
            class="TabFocusTrap__TabFocusTrap___39Vty"
            tabindex="-1"
          >
            Give your feedback
          </span>
        </button>
      </div>
    </div>
    <div
      class="Workbench__Workbench__content-wrapper___3yWKi"
    >
      <div
        class="Workbench__Workbench__content___1U8bV"
        data-test-id="cf-ui-workbench-content"
      >
        <div
          class="Workbench__Workbench__content-inner___36_Hx Workbench__Workbench__content-inner--text___7nfTX"
        >
          <div
            class="Card__Card___xXJsT emotion-4 Card__Card--padding-large___3Fgal"
            data-test-id="cf-ui-card"
          >
            <div>
              <h1
                class="Heading__Heading___83D3K emotion-0"
                data-test-id="cf-ui-heading"
              >
                Introducing Contentful Apps
              </h1>
              <p
                class="Paragraph__Paragraph___2aO-9 emotion-1"
                data-test-id="cf-ui-paragraph"
              >
                Contentful apps extend the capabilities of the Contentful web app and the editors who use it. Apps empower you to integrate third-party services, build efficient workflows and customize the functionality of the Contentful web app.
              </p>
              <a
                class="Button__Button___1ZfFj a11y__focus-border--default___60AXp emotion-2 Button__Button--primary___JImeO"
                data-test-id="cf-ui-button"
                href="https://www.contentful.com/developers/docs/extensibility/app-framework/"
                rel="noopener noreferrer"
                target="_blank"
                type="button"
              >
                <span
                  class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
                  tabindex="-1"
                >
                  <span
                    class="Button__Button__label___3tcOj"
                  >
                    Learn how to build your own app
                  </span>
                </span>
              </a>
            </div>
            <div
              class="emotion-3"
            >
              <connected-shapes.svg
                classname="css-1z01nmi-svgContainerExtension"
              />
            </div>
          </div>
          <div
            class="Card__Card___xXJsT emotion-48 Card__Card--padding-large___3Fgal"
            data-test-id="cf-ui-card"
          >
            <p
              class="Paragraph__Paragraph___2aO-9 emotion-5"
              data-test-id="cf-ui-paragraph"
            >
              Contentful apps extend and expand the capabilities of the Contentful web app. You can integrate your favorite third-party services, build better workflows and customize what you can do with Contentful.
            </p>
            <div>
              <hr />
              <h2
                class="Heading__Heading___83D3K"
                data-test-id="cf-ui-heading"
              >
                Available
              </h2>
              <div>
                <div
                  class="emotion-12"
                >
                  <div
                    class="emotion-10"
                    data-test-id="app-title"
                  >
                    <h3
                      class="Heading__Heading___83D3K emotion-9"
                      data-test-id="cf-ui-heading"
                    >
                      <div
                        class="emotion-8"
                        data-test-id="app-details"
                      >
                        <img
                          class="emotion-6"
                          src="//images.ctfassets.net/lpjm8d10rkpy/2dYQ3x6VMNZJgDXYmMzLBx/b338ad4b50cd0ff31e7157671b801b4d/logomark.svg"
                        />
                        <div>
                          Netlify
                          <div
                            class="emotion-7"
                          >
                            Easily build and preview sites with the Netlify app 
                          </div>
                        </div>
                      </div>
                    </h3>
                  </div>
                  <div
                    class="emotion-11"
                  >
                    <button
                      class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                      data-test-id="cf-ui-text-link"
                      type="button"
                    >
                      <span
                        class="TabFocusTrap__TabFocusTrap___39Vty"
                        tabindex="-1"
                      >
                        About
                      </span>
                    </button>
                  </div>
                </div>
                <div
                  class="emotion-12"
                >
                  <div
                    class="emotion-10"
                    data-test-id="app-title"
                  >
                    <h3
                      class="Heading__Heading___83D3K emotion-9"
                      data-test-id="cf-ui-heading"
                    >
                      <div
                        class="emotion-8"
                        data-test-id="app-details"
                      >
                        <img
                          class="emotion-6"
                          src="//images.ctfassets.net/lpjm8d10rkpy/4X7O4Q0pIgQZNcONoQrQlp/9262ad9a935fa92e9cacd9207ae0a401/optimizely-logo.svg"
                        />
                        <div>
                          Optimizely
                          <div
                            class="emotion-7"
                          >
                            Enable experimentation at scale with the Optimizely app
                          </div>
                        </div>
                      </div>
                    </h3>
                  </div>
                  <div
                    class="emotion-11"
                  >
                    <button
                      class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                      data-test-id="cf-ui-text-link"
                      type="button"
                    >
                      <span
                        class="TabFocusTrap__TabFocusTrap___39Vty"
                        tabindex="-1"
                      >
                        About
                      </span>
                    </button>
                  </div>
                </div>
                <div
                  class="emotion-12"
                >
                  <div
                    class="emotion-10"
                    data-test-id="app-title"
                  >
                    <h3
                      class="Heading__Heading___83D3K emotion-9"
                      data-test-id="cf-ui-heading"
                    >
                      <div
                        class="emotion-8"
                        data-test-id="app-details"
                      >
                        <img
                          class="emotion-6"
                          src="//images.ctfassets.net/lpjm8d10rkpy/1o6HKvOxxumeW64hv6ANMw/b1c5308bcd4e48cfaafc185c0b12aa36/Gatsby_Monogram.svg"
                        />
                        <div>
                          Gatsby Cloud
                          <div
                            class="emotion-7"
                          >
                            Instantly see content changes on your website with the Gatsby app
                          </div>
                        </div>
                      </div>
                    </h3>
                  </div>
                  <div
                    class="emotion-11"
                  >
                    <button
                      class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                      data-test-id="cf-ui-text-link"
                      type="button"
                    >
                      <span
                        class="TabFocusTrap__TabFocusTrap___39Vty"
                        tabindex="-1"
                      >
                        About
                      </span>
                    </button>
                  </div>
                </div>
                <div
                  class="emotion-12"
                >
                  <div
                    class="emotion-10"
                    data-test-id="app-title"
                  >
                    <h3
                      class="Heading__Heading___83D3K emotion-9"
                      data-test-id="cf-ui-heading"
                    >
                      <div
                        class="emotion-8"
                        data-test-id="app-details"
                      >
                        <img
                          class="emotion-6"
                          src="//images.ctfassets.net/lpjm8d10rkpy/3ZH89y9VHHd53v4OQ3BG0V/cc62043a88cd5fb80719687f299e7351/cloudinary_icon_for_white_bg.svg"
                        />
                        <div>
                          Cloudinary
                          <div
                            class="emotion-7"
                          >
                            Seamlessly deliver images and videos with the Cloudinary app
                          </div>
                        </div>
                      </div>
                    </h3>
                  </div>
                  <div
                    class="emotion-11"
                  >
                    <button
                      class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                      data-test-id="cf-ui-text-link"
                      type="button"
                    >
                      <span
                        class="TabFocusTrap__TabFocusTrap___39Vty"
                        tabindex="-1"
                      >
                        About
                      </span>
                    </button>
                  </div>
                </div>
                <div
                  class="emotion-12"
                >
                  <div
                    class="emotion-10"
                    data-test-id="app-title"
                  >
                    <h3
                      class="Heading__Heading___83D3K emotion-9"
                      data-test-id="cf-ui-heading"
                    >
                      <div
                        class="emotion-8"
                        data-test-id="app-details"
                      >
                        <img
                          class="emotion-6"
                          src="//images.ctfassets.net/lpjm8d10rkpy/3vNJZRjthatucy2mBJMy1u/0e5d501207427780021e7580cc384761/bynder.5e9e457b9d.svg"
                        />
                        <div>
                          Bynder
                          <div
                            class="emotion-7"
                          >
                            Find and use brand, marketing and product assets from your Bynder library 
                          </div>
                        </div>
                      </div>
                    </h3>
                  </div>
                  <div
                    class="emotion-11"
                  >
                    <button
                      class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                      data-test-id="cf-ui-text-link"
                      type="button"
                    >
                      <span
                        class="TabFocusTrap__TabFocusTrap___39Vty"
                        tabindex="-1"
                      >
                        About
                      </span>
                    </button>
                  </div>
                </div>
                <div
                  class="emotion-12"
                >
                  <div
                    class="emotion-10"
                    data-test-id="app-title"
                  >
                    <h3
                      class="Heading__Heading___83D3K emotion-9"
                      data-test-id="cf-ui-heading"
                    >
                      <div
                        class="emotion-8"
                        data-test-id="app-details"
                      >
                        <img
                          class="emotion-6"
                          src="//images.ctfassets.net/lpjm8d10rkpy/7KdyOX9FlBz0SX5QyaLFaF/6ee4f786204d6f80b096fb73584dbcf7/logo.svg"
                        />
                        <div>
                          Dropbox
                          <div
                            class="emotion-7"
                          >
                            Connect Contentful and Dropbox to easily share content 
                          </div>
                        </div>
                      </div>
                    </h3>
                  </div>
                  <div
                    class="emotion-11"
                  >
                    <button
                      class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
                      data-test-id="cf-ui-text-link"
                      type="button"
                    >
                      <span
                        class="TabFocusTrap__TabFocusTrap___39Vty"
                        tabindex="-1"
                      >
                        About
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
